[Docs] ES|QL Data Federation (#146945)
elastic/elasticsearch: docs/reference/query-languages/esql/esql-data-federation-sources.md

| Model | `auth` value | Description |
|---|---|---|
| Static credentials | `static_credentials` | A fixed access key and secret key, optionally with a session token for temporary credentials. The common form for a service account. To set one up, refer to [connect with static credentials](esql-data-federation-static-credentials.md). |
| Anonymous | `anonymous` | For public data that needs no credentials. The [quickstart](esql-data-federation-quickstart.md) walks through this method. |
| Federated identity | `federated_identity` | Keyless. {{es}} exchanges a short-lived OIDC token for temporary AWS credentials via STS, so no static keys are stored. Available on Elastic Cloud Hosted and serverless only. Operator-gated (`esql.datasource.federated_identity.enabled`). To set it up, refer to [connect with federated identity](esql-data-federation-federated-identity.md). |
| Managed identity | `managed_identity` | Keyless. Uses the {{es}} node's own cloud identity, for example an EC2 instance IAM role. Operator-only and API-only, and not available in serverless. Requires `esql.datasource.managed_identity.enabled`. |

> | Auto (API-only) | `auto` (default) | Infers the authentication mode from the settings you provide. When you supply no keys, it falls back to the default AWS credential chain (IAM roles, environment variables, or instance profiles). API-only. |

Reviewer: Do we really need to document that? It's really bad practice and we should keep it for internal/testing use only. Since this is not an API spec, I'd rather remove it.

Author: Removing, thanks!

> | Auto (API-only) | `auto` (default) | Infers the authentication mode from the settings you provide. When you supply no keys, it falls back to the default AWS credential chain (IAM roles, environment variables, or instance profiles). API-only. |

Author: ```suggestion
```

:::{image} images/data-federation/data-sources-tab.png
:alt: The Data sources tab listing several registered Amazon S3 data sources with their dataset counts, descriptions, and edit and delete row actions
:width: 800px
:::

### Connect a new data source

Click **Connect data source** to open a flyout where you define the connection:

- **Data source type**: the storage system to connect to, such as **Amazon S3**.
- **Name**: a unique name for the data source.
- **Description**: an optional description.
- **Region**: the storage region.
- **Endpoint**: an optional Amazon S3 endpoint override.
- **Authentication**: a **Preferred method** and its credentials. For example, **Access and Secret Keys** takes an **Access key** and **Secret key**.

For the full set of authentication methods, refer to [authentication models](#authentication).

:::{dropdown} Show the Connect data source flyout
:::{image} images/data-federation/connect-data-source-static-credentials.png
:alt: The Connect external data source flyout for an Amazon S3 data source, with the Access and Secret Keys authentication method selected
:width: 450px
:::
:::

## Manage data sources using the API

> Data sources are managed under the `/_query/data_source` endpoint. All data source operations require the cluster `manage` privilege or a `global.data_source` privilege (refer to [manage credentials and privileges](esql-data-federation-security.md)).

Reviewer 2: Link to the API docs for the endpoint?

Author: Resolved in https://github.com/elastic/elasticsearch/commit/48b0b73a4a5d1d2a5af7a829bbb3963c0582f9b2

| Operation | Endpoint |
|---|---|
| [Create or update](#create-or-update-a-data-source) | `PUT /_query/data_source/{name}` |
| [Get](#get-a-data-source) | `GET /_query/data_source/{name}` |
| [List all](#list-all-data-sources) | `GET /_query/data_source` |
| [Delete](#delete-a-data-source) | `DELETE /_query/data_source/{name}` |

<!-- # https://www.elastic.co/docs/api/doc/elasticsearch/operation/operation-esql-put-data-source -->
<!-- # https://www.elastic.co/docs/api/doc/elasticsearch/operation/operation-esql-get-data-source -->
<!-- # https://www.elastic.co/docs/api/doc/elasticsearch/operation/operation-esql-delete-data-source -->

### Create or update a data source

> `PUT` creates a new data source or replaces an existing one entirely. The create request does not validate connectivity to the external system. To verify that credentials and endpoint are correct, create a dataset that references the data source and query it.

Reviewer 3: Three things here: an omitted secret is carried forward (not cleared) on update; the name rules aren't stated; and the 100-data-source cap bites at create.

```suggestion
`PUT` creates a new data source or replaces an existing one entirely with one exception. Secrets you omit from the request are carried forward from the existing definition rather than cleared, so you can update non-secret settings without re-sending credentials. The create request does not validate connectivity to the external system. To verify that credentials and endpoint are correct, create a dataset that references the data source and query it.

:::{important}
Data source names follow the same naming rules as index names: lowercase only, at most 255 bytes, and they cannot begin with `-`, `_`, or `+`, contain spaces, or contain the characters `\ / * ? " < > |`.
:::

A cluster holds at most 100 data sources by default. Creating one past the limit fails; raise it with the [`esql.data_sources.max_count`](esql-data-federation-cluster-settings.md#object-limits) cluster setting.
```

::::{tab-set}
:group: api-ref

:::{tab-item} Console
:sync: console
```console
PUT /_query/data_source/prod_s3_logs
{
  "type": "s3",
  "description": "Production S3 logs bucket, us-east-1",
  "settings": {
    "region": "us-east-1",
    "auth": "static_credentials",
    "access_key": "<AWS_ACCESS_KEY_ID>",
    "secret_key": "<AWS_SECRET_ACCESS_KEY>"
  }
}
```
:::

:::{tab-item} curl
:sync: curl
```bash
curl -X PUT "${ELASTICSEARCH_URL}/_query/data_source/prod_s3_logs" \
  -H "Authorization: ApiKey ${API_KEY}" \
  -H "Content-Type: application/json" \
  -d '{
  "type": "s3",
  "description": "Production S3 logs bucket, us-east-1",
  "settings": {
    "region": "us-east-1",
    "auth": "static_credentials",
    "access_key": "<AWS_ACCESS_KEY_ID>",
    "secret_key": "<AWS_SECRET_ACCESS_KEY>"
  }
}'
```
:::

::::

### Get a data source

> Retrieves a data source by name. Credential values are replaced by `::es_redacted::` in the response.

Reviewer 3: GET also accepts comma-separated names and `*` wildcards.

```suggestion
Retrieves a data source by name. You can pass a comma-separated list of names and use `*` wildcards. A concrete name that does not exist returns a `404`; a wildcard that matches nothing returns an empty list. Credential values are replaced by `::es_redacted::` in the response.
```

::::{tab-set}
:group: api-ref

:::{tab-item} Console
:sync: console
```console
GET /_query/data_source/prod_s3_logs
```
:::

:::{tab-item} curl
:sync: curl
```bash
curl -X GET "${ELASTICSEARCH_URL}/_query/data_source/prod_s3_logs" \
  -H "Authorization: ApiKey ${API_KEY}"
```
:::

::::

### List all data sources

Returns all registered data sources.

::::{tab-set}
:group: api-ref

:::{tab-item} Console
:sync: console
```console
GET /_query/data_source
```
:::

:::{tab-item} curl
:sync: curl
```bash
curl -X GET "${ELASTICSEARCH_URL}/_query/data_source" \
  -H "Authorization: ApiKey ${API_KEY}"
```
:::

::::

### Delete a data source

> Deletes a data source by name.

Reviewer 3: DELETE accepts a comma-separated list and is all-or-nothing (404 if any name is missing).

```suggestion
Deletes one or more data sources by name. You can pass a comma-separated list. The request is all-or-nothing: if any named data source does not exist, the request returns a `404` and nothing is deleted.
```

::::{tab-set}
:group: api-ref

:::{tab-item} Console
:sync: console
```console
DELETE /_query/data_source/prod_s3_logs
```
:::

:::{tab-item} curl
:sync: curl
```bash
curl -X DELETE "${ELASTICSEARCH_URL}/_query/data_source/prod_s3_logs" \
  -H "Authorization: ApiKey ${API_KEY}"
```
:::

::::

:::{important}
A data source cannot be deleted while datasets still reference it. Delete the dependent datasets first, or the request returns a `409 Conflict` error.
:::

## Data source settings

Settings vary by data source type.

### S3

The following settings are available for `s3` data sources:

**Connection settings:**

| Setting | Required | Description |
|---|---|---|
| `region` | No | The bucket's AWS region, for example `us-east-1`. Defaults to `us-east-1` if omitted. Set it to match the bucket's region, otherwise requests to the bucket fail. |

> | `endpoint` | No | An explicit Amazon S3 endpoint override. |

Reviewer 3: Setting `endpoint` also forces path-style addressing.

```suggestion
| `endpoint` | No | An explicit Amazon S3 endpoint override. Setting it switches requests to path-style addressing. |
```

:::{tip}
A data source connects to a single region. To query buckets in more than one region, create a separate data source for each region.
:::

**Authentication settings:**

| Setting | Required | Description |
|---|---|---|
| `access_key` | No | AWS access key ID. Used with `auth: static_credentials`. |
| `secret_key` | No | AWS secret access key. Used with `auth: static_credentials`. |

> | `session_token` | No | Session token, when using temporary credentials. Use with `access_key` and `secret_key`. |

Reviewer 3: The **Connect data source** flyout doesn't expose session tokens, so let's not surface `session_token` in the docs for the preview either. It's supported, but I'd drop the row for now to keep the docs and the UI in step.

```suggestion

```

| `role_arn` | Yes (federated identity) | The ARN of the IAM role {{es}} assumes via STS. Used with `auth: federated_identity`. |

> | `jwt_audience` | No | Overrides the JWT audience claim sent to STS. Used with `auth: federated_identity`. |

Reviewer 3: Default is `sts.amazonaws.com`.

```suggestion
| `jwt_audience` | No | Overrides the JWT audience claim sent to STS. Defaults to `sts.amazonaws.com`. Used with `auth: federated_identity`. |
```

> | `role_session_name` | No | A label for the assumed-role session. Used with `auth: federated_identity`. |

Reviewer 3: Default is `elasticsearch-esql-datasource`.

```suggestion
| `role_session_name` | No | A label for the assumed-role session. Defaults to `elasticsearch-esql-datasource`. Used with `auth: federated_identity`. |
```

| `sts_endpoint` | No | A custom STS endpoint URL. Used with `auth: federated_identity`. |

> | `sts_region` | No | The AWS region of the STS endpoint. Used with `auth: federated_identity`. |

Reviewer 3: Defaults to the bucket's region.

```suggestion
| `sts_region` | No | The AWS region of the STS endpoint. Defaults to the bucket's region. Used with `auth: federated_identity`. |
```

> | `auth` | Yes | Authentication mode. Set it to `anonymous`, `static_credentials`, `managed_identity`, or `federated_identity`. |

Reviewer 3: `auth` isn't required — omitting it selects `auto`, which infers the mode from the credentials present. The value list is also missing `auto`.

```suggestion
| `auth` | No | Authentication mode. Valid values: `auto` (the default), `anonymous`, `static_credentials`, `federated_identity`, `managed_identity`. When omitted, `auth` defaults to `auto` and {{es}} infers the mode from the credentials you supply. |
```

Author: let's check with @bytebilly  for consensus on this before publication, the omission was intentional 👍

## Authentication

> A data source authenticates to its store with one of the models below. The models are mutually exclusive on a data source.

Reviewer 4: ```suggestion
A data source authenticates to its store with one of the following models. The models are mutually exclusive on a data source.
```

> A data source authenticates to its store with one of the following models. The models are mutually exclusive on a data source.

Reviewer 3: Worth spelling out what `auto` does, since it's the default and drives the whole table.

```suggestion
A data source authenticates to its store with one of the following models. The models are mutually exclusive on a data source.
Setting `auth` explicitly is optional. When omitted, it defaults to `auto`, and {{es}} infers the mode from the credentials you supply: `static_credentials` when `access_key` and `secret_key` are present, or `federated_identity` when the federated identity settings are present. `auto` never infers `anonymous` or `managed_identity` — set those explicitly. A request with no credentials and no explicit `auth` is rejected.
```

Author: same comment as previous

| Model | `auth` value | Description |
|---|---|---|

> | Static credentials | `static_credentials` | A fixed access key and secret key, optionally with a session token for temporary credentials. The common form for a service account. To set one up, refer to [connect with static credentials](esql-data-federation-static-credentials.md). |

Reviewer 3: Dropping the session-token mention here too, to match the settings table.

```suggestion
| Static credentials | `static_credentials` | A fixed access key and secret key. The common form for a service account. To set one up, refer to [connect with static credentials](esql-data-federation-static-credentials.md). |
```

| Anonymous | `anonymous` | For public data that needs no credentials. The [quickstart](esql-data-federation-quickstart.md) walks through this method. |
| Federated identity | `federated_identity` | Keyless. {{es}} exchanges a short-lived OIDC token for temporary AWS credentials via STS, so no static keys are stored. Available on Elastic Cloud Hosted and serverless only. Operator-gated (`esql.datasource.federated_identity.enabled`). To set it up, refer to [connect with federated identity](esql-data-federation-federated-identity.md). |

> | Managed identity | `managed_identity` | Keyless. Uses the {{es}} node's own cloud identity, for example an EC2 instance IAM role. Operator-only and API-only, and not available in serverless. Requires `esql.datasource.managed_identity.enabled`. |

Reviewer 3: The old value names still work — worth noting so existing configs aren't surprising.

```suggestion
| Managed identity | `managed_identity` | Keyless. Uses the {{es}} node's own cloud identity, for example an EC2 instance IAM role. Operator-only and API-only, and not available in serverless. Requires `esql.datasource.managed_identity.enabled`. |
:::{note}
`none` and `workload_identity` are deprecated aliases for `anonymous` and `managed_identity`. They are still accepted, with a deprecation warning, and are normalized to the canonical values.
:::
```

Author: Again, I think we can safely omit this, keeping comment open for visibility

---
navigation_title: "Connect data sources"
description: "Connect Elasticsearch to external storage with ES|QL Data Federation by setting up S3 data sources, configuring regions and endpoints, and authenticating access."
applies_to:
  stack: preview =9.5
  serverless: preview
products:
  - id: elasticsearch
---

# Connect external data sources for {{esql}} Data Federation

A data source defines the connection to an external storage system. It stores the connection type, region, endpoint, and credentials. A data source defines how to connect, not what data to query. One data source can serve many [datasets](esql-data-federation-datasets.md). When credentials rotate, you update the data source in one place without touching the datasets that reference it.

## Supported data source types

The following data source types are supported:

:::{include} _snippets/data-federation/supported-data-source-types.md
:::

:::{note}
Other S3-compatible services have not been validated and are not supported for this technical preview.
:::

> ## Manage data sources in the UI

Reviewer 5: wonder if we perhaps prefer "using Kibana" in these scenarios

Author: Fixed in https://github.com/elastic/elasticsearch/pull/146945/commits/affeef8e8fee

In {{kib}}, you connect and manage data sources from the **Data sources** tab under **Data management** > **{{esql}} Data Federation**.

The **Data sources** tab lists each registered data source including:
-  its type
-  its description
-  the number of datasets that reference it

From this tab you can search your data sources, connect a new one, and edit or delete an existing one.

:::{image} images/data-federation/data-sources-tab.png
:alt: The Data sources tab listing several registered Amazon S3 data sources with their dataset counts, descriptions, and edit and delete row actions
:width: 800px
:::

### Connect a new data source

Click **Connect data source** to open a flyout where you define the connection:

- **Data source type**: the storage system to connect to, such as **Amazon S3**.
- **Name**: a unique name for the data source.
- **Description**: an optional description.

> - **Region**: the storage region.

Reviewer 5: same with this one. could probably be a little more verbose/clear here

Author: Fixed in https://github.com/leemthompo/elasticsearch/commit/d54c5c5cbd14d1c7f48b53b3c049f409ece175dc

- **Endpoint**: an optional Amazon S3 endpoint override.

> - **Authentication**: a **Preferred method** and its credentials. For example, **Access and Secret Keys** takes an **Access key** and **Secret key**.

Reviewer 5: what does preferred method mean here? could use a slightly more verbose explananation ... this just restates the field name

Author: Fixed in https://github.com/leemthompo/elasticsearch/commit/d54c5c5cbd14d1c7f48b53b3c049f409ece175dc

> For the full set of authentication methods, refer to [authentication models](#authentication).

Reviewer 5: maybe hinting that we have deeper-dive tutorials for this would be helpful. "For information about setting up different authentication models in AWS..." or something

Author: Fixed in https://github.com/leemthompo/elasticsearch/commit/d54c5c5cbd14d1c7f48b53b3c049f409ece175dc

:::{dropdown} Show the Connect data source flyout
:::{image} images/data-federation/connect-data-source-static-credentials.png
:alt: The Connect external data source flyout for an Amazon S3 data source, with the Access and Secret Keys authentication method selected
:width: 450px
:::
:::

## Manage data sources using the API

<!-- TODO: link to API reference once https://github.com/elastic/elasticsearch-specification/pull/6503 is published -->

> Data sources are managed under the `/_query/data_source` endpoint. All data source operations require the cluster `manage` privilege or a `global.data_source` privilege (refer to [manage credentials and privileges](esql-data-federation-security.md)).

Reviewer 5: would avoid parentheticals like this

Author: Fixed in https://github.com/elastic/elasticsearch/pull/146945/commits/affeef8e8fee

| Operation | Endpoint |
|---|---|
| [Create or update](#create-or-update-a-data-source) | `PUT /_query/data_source/{name}` |
| [Get](#get-a-data-source) | `GET /_query/data_source/{name}` |
| [List all](#list-all-data-sources) | `GET /_query/data_source` |
| [Delete](#delete-a-data-source) | `DELETE /_query/data_source/{name}` |

<!-- # https://www.elastic.co/docs/api/doc/elasticsearch/operation/operation-esql-put-data-source -->
<!-- # https://www.elastic.co/docs/api/doc/elasticsearch/operation/operation-esql-get-data-source -->
<!-- # https://www.elastic.co/docs/api/doc/elasticsearch/operation/operation-esql-delete-data-source -->

### Create or update a data source

> `PUT` creates a new data source or replaces an existing one entirely with one exception. Secrets you omit from the request are carried forward from the existing definition rather than cleared, so you can update non-secret settings without re-sending credentials. The create request does not validate connectivity to the external system. To verify that credentials and endpoint are correct, create a dataset that references the data source and query it.

Reviewer 5: ```suggestion
`PUT` creates a new data source or replaces an existing one entirely with one exception. Secrets you omit from the request are carried forward from the existing definition rather than cleared, so you can update non-secret settings without re-sending credentials. 

The create request does not validate connectivity to the external system. To verify that credentials and endpoint are correct, create a dataset that references the data source and query it.
```

Author: Fixed in https://github.com/elastic/elasticsearch/pull/146945/commits/118f93d4f9a0

:::{important}
Data source names follow the same naming rules as index names: lowercase only, at most 255 bytes, and they cannot begin with `-`, `_`, or `+`, contain spaces, or contain the characters `\ / * ? " < > |`.
:::

> A cluster holds at most 100 data sources by default. Creating one past the limit fails; raise it with the [`esql.data_sources.max_count`](esql-data-federation-cluster-settings.md#object-limits) cluster setting.

Reviewer 5: this brevity feels like a hallmark of AI. we have lots of room here to be more explicit.

I don't think we need to say "creating one past the limit fails" - "at most" implies that creation after that point is not allowed

stray thought: serverless users are stranded here because they can't use these settings. not sure what else in these docs has the same problem

```suggestion
A cluster holds at most 100 data sources by default. In {{stack}} deployments, if you need more than 100 data sources, then you can raise the limit using the [`esql.data_sources.max_count`](esql-data-federation-cluster-settings.md#object-limits) cluster setting.
```

Author: Fixed in https://github.com/elastic/elasticsearch/pull/146945/commits/118f93d4f9a0

> A cluster holds at most 100 data sources by default. Creating one past the limit fails; raise it with the [`esql.data_sources.max_count`](esql-data-federation-cluster-settings.md#object-limits) cluster setting.

Reviewer 5: this limitation probably also holds true for the UI, and should be moved up ... same with the naming stuff.

Author: Fixed in https://github.com/leemthompo/elasticsearch/commit/d54c5c5cbd14d1c7f48b53b3c049f409ece175dc

::::{tab-set}
:group: api-ref

:::{tab-item} Console
:sync: console
```console
PUT /_query/data_source/prod_s3_logs
{
  "type": "s3",
  "description": "Production S3 logs bucket, us-east-1",
  "settings": {
    "region": "us-east-1",
    "auth": "static_credentials",
    "access_key": "<AWS_ACCESS_KEY_ID>",

>     "secret_key": "<AWS_SECRET_ACCESS_KEY>"

Reviewer 5: might want some footnotes here saying we tell you how to do this in detail on other pages?

Author: Fixed in https://github.com/leemthompo/elasticsearch/commit/d54c5c5cbd14d1c7f48b53b3c049f409ece175dc

  "type": "s3",
  "description": "Production S3 logs bucket, us-east-1",
  "settings": {
    "region": "us-east-1",
    "auth": "static_credentials",
    "access_key": "<AWS_ACCESS_KEY_ID>",
    "secret_key": "<AWS_SECRET_ACCESS_KEY>"
  }
}'
```
:::

::::

### Get a data source

Retrieves a data source by name. You can pass a comma-separated list of names and use `*` wildcards. A concrete name that does not exist returns a `404`; a wildcard that matches nothing returns an empty list. Credential values are replaced by `::es_redacted::` in the response.

::::{tab-set}
:group: api-ref

:::{tab-item} Console
:sync: console
```console
GET /_query/data_source/prod_s3_logs
```
:::

:::{tab-item} curl
:sync: curl
```bash
curl -X GET "${ELASTICSEARCH_URL}/_query/data_source/prod_s3_logs" \
  -H "Authorization: ApiKey ${API_KEY}"
```
:::

::::

### List all data sources

Returns all registered data sources.

::::{tab-set}
:group: api-ref

:::{tab-item} Console
:sync: console
```console
GET /_query/data_source
```
:::

:::{tab-item} curl
:sync: curl
```bash
curl -X GET "${ELASTICSEARCH_URL}/_query/data_source" \
  -H "Authorization: ApiKey ${API_KEY}"
```
:::

::::

### Delete a data source

> Deletes one or more data sources by name. You can pass a comma-separated list. The request is all-or-nothing: if any named data source does not exist, the request returns a `404` and nothing is deleted.

Reviewer 5: this is a little alarming. we're just talking about an atomic request. prob can skip the jargon

```suggestion
Deletes one or more data sources by name. You can pass a comma-separated list. If any named data source does not exist, the request returns a `404` and nothing is deleted.
```

Author: Fixed in https://github.com/elastic/elasticsearch/pull/146945/commits/118f93d4f9a0

::::{tab-set}
:group: api-ref

:::{tab-item} Console
:sync: console
```console
DELETE /_query/data_source/prod_s3_logs
```
:::

:::{tab-item} curl
:sync: curl
```bash
curl -X DELETE "${ELASTICSEARCH_URL}/_query/data_source/prod_s3_logs" \
  -H "Authorization: ApiKey ${API_KEY}"
```
:::

::::

:::{important}
A data source cannot be deleted while datasets still reference it. Delete the dependent datasets first, or the request returns a `409 Conflict` error.
:::

> ## Data source settings

Reviewer 5: same note about settings ... possibly need different treatment (a tiny bit less sure for this one, but feels similar to how we handle backup and restore credentials, which are ultimately cluster settings)

Author: Noted, will keep this in mind for a follow-up pass.
